pkg/scheduler/framework/v1alpha1/interface.go

```diff
@@ -273,6 +280,20 @@ type FilterPlugin interface {
 	Filter(ctx context.Context, state *CycleState, pod *v1.Pod, nodeInfo *NodeInfo) *Status
 }
 
+// PostFilterPlugin is an interface for PostFilter plugins. These plugins are called
+// after a pod cannot be scheduled.
+type PostFilterPlugin interface {
+	Plugin
+	// PostFilter is called by the scheduling framework.
+	// A PostFilter should return one of the following statuses:
+	// - Unschedulable: a plugin ran its logic but still cannot make it schedulable.
+	// - Schedulable: a plugin ran its logic and can make it schedulable.
+	// - Error: some internal error occurs.
+	// - Success: indicate the plugin ran its logic successfully, but it has no idea whether
+	//   the pod is schedulable or not.
+	PostFilter(ctx context.Context, state *CycleState, pod *v1.Pod, filterErr error, ph PreemptHandle) (string, *Status)
```

Comment:
Re 1: FitError has wrapped `framework.NodeToStatusMap`.

Re 2: I discussed this with Aldo. Generally we think FrameworkHandle is more a generic interface which may be used in most extension points. While for PreemptHandle, it should be wired to PostFilter extension point only - as there is no example proving that preemption is needed in scheduling phases other than PostFilter.

Reply:
> While for PreemptHandle, it should be wired to PostFilter extension point only

+1

Reply:
> Re 1: FitError has wrapped framework.NodeToStatusMap.

yeah, I noticed that, but I see no value to the FitError wrapper. The old PostFilter interface used to pass both the filtered list and the statuses map, I think that is more generic. We don't need a FitError type at all. A PostFilter plugin implementing preemption logic can simply check ```len(filteredNodes) == 0``` to whether or not to execute.

> Re 2: I discussed this with Aldo. Generally we think FrameworkHandle is more a ...

I think things we pass through the Plugin function call should have a limited scope, usually the cycle scope, this also simplifies the function signature. Things like FrameworkHandle and PreemptHandle have cross-cycle scope, we want to emphasize that by not passing it in each invocation. And I come back to the argument that PostFilter is a generic extension point.

Reply:
> yeah, I noticed that, but I see no value to the FitError wrapper. The old PostFilter interface used to pass both the filtered list and the statuses map, I think that is more generic. We don't need a FitError type at all. A PostFilter plugin implementing preemption logic can simply check len(filteredNodes) == 0 to whether or not to execute.

SG. Later we may be able to eliminate FitError.

> I think things we pass through the Plugin function call should have a limited scope, usually the cycle scope, this also simplifies the function signature. Things like FrameworkHandle and PreemptHandle have cross-cycle scope, we want to emphasize that by not passing it in each invocation. And I come back to the argument that PostFilter is a generic extension point.

It's bit debatable as both options have their pros and cons. But you're right, it's important to be aligned with the design rationale that individual extension point should be exposing data/struct related with that scheduling cycle. Maybe it will pay us back more.

@everpeace @alculquicondor @denkensk WDYT?

Reply:
> A PostFilter plugin implementing preemption logic can simply check len(filteredNodes) == 0 to whether or not to execute.

Now we are saying that PostFilter plugins can run even if Filtering failed to find candidates. I'm ok with that, but let's agree. (I suggest we sync at today's meeting).

I think, in the future, we should have a Handle interface for each extension point (with a unique underlying implementation). The idea is that we limit actions that can be taken in each EP. Aligning with the idea that PostFilter is generic, the Handle should be named PostFilterHandle.

However, given that we are not doing that in any other EP today, we can just leave it in the FrameworkHandle. But, for the purpose of separation of concerns, we can still have a separate interface for PreemptionHandle that can be obtained from the FrameworkHandle.

Reply:
> Now we are saying that PostFilter plugins can run even if Filtering failed to find candidates. I'm ok with that, but let's agree. (I suggest we sync at today's meeting).

PostFilter will run only when filtering fails to find a node, so passing filteredNodes means nothing now. The only thing we need to pass is the failedStatusesMap.

> I think, in the future, we should have a Handle interface for each extension point (with a unique underlying implementation). The idea is that we limit actions that can be taken in each EP. Aligning with the idea that PostFilter is generic, the Handle should be named PostFilterHandle.

I think this adds too much burden on the framework maintainers (deciding what should and shouldn't be included in each "handle").